### DISCUSSION TITLE
What is the motivation behind the widget class api?

### DISCUSSION THREAD

**Discussion by fetishcoder** [visgl/deck.gl#8110] [General]
Hi guys. Let me start by saying I love this lib so my question comes from a good place :)

What is the motivation behind the widget class api?
This clearly seems to represent some component.
So why not use WebComponents? ( its a standard since 2018 and you don't _have_ to use the ShadowDom)
Why introduce your own api for render, click, drag etc..?
Seems like more code to maintain?
Also, why put the manager in core?
Deck.gl has always been remarkably clean in their separation of concerns!

Sorry to bother you with these questions, but I do not know the place where you are discussing these things.

Thanks for reading!

**Comment by Pessimistress** [comment]
Thanks for taking an interest in the widgets effort.
You can find the RFC at https://github.com/visgl/deck.gl/blob/master/dev-docs/RFCs/v8.0/ui-module-rfc.md

> Why introduce your own api for render, click, drag etc..?
> Also, why put the manager in core?

Because these are events that happen inside the Deck instance - "render" refers to when Deck redraws the WebGL context; "click" refers to when an object in a layer is clicked; etc. So yes, these are our own APIs, and they can only be supported if the core is aware of the widgets mechanism.

> why not use WebComponents?

I have nothing against WebComponents, just not clear how it will benefit us.

**Comment by fetishcoder** [comment]
Thank you

**Comment by ibgreen** [comment]
It seems to me that Web Components could perhaps still be an option for how to package up these widgets and make it easy to use them in all web apps. We do want something that is "standard", works with pure javascript and independent of framework (React etc). @chrisgervang FYI.
